File: dotnet/Falcon.PluginAPI/Items/Item.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class Item : IDisposable
    {
        private FalconItem* _handle;
        private bool _owned;

        private Item(FalconItem* handle, bool owned)
        {
            _handle = handle;
            _owned = owned;
        }

        public nint Handle
        {
            get
            {
                return (nint)_handle;
            }
        }

        public bool IsValid
        {
            get
            {
                return _handle != null;
            }
        }

        public bool IsOwned
        {
            get
            {
                return _owned;
            }
        }

        public bool IsEmpty
        {
            get
            {
                return _handle == null || NativeApi.Table->itemIsEmpty(_handle) != 0;
            }
        }

        public string Identifier
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->itemIdentifier(Raw));
            }
        }

        public uint Count
        {
            get
            {
                return NativeApi.Table->itemCount(Raw);
            }
            set
            {
                NativeApi.Table->itemSetCount(Raw, value);
            }
        }

        public int Damage
        {
            get
            {
                return NativeApi.Table->itemDamage(Raw);
            }
            set
            {
                NativeApi.Table->itemSetDamage(Raw, value);
            }
        }

        public string CustomName
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->itemCustomName(Raw));
            }
            set
            {
                fixed (byte* name = NativeApi.Utf8(value))
                {
                    NativeApi.Table->itemSetCustomName(Raw, name);
                }
            }
        }

        public IReadOnlyList<string> Lore
        {
            get
            {
                FalconItem* handle = Raw;
                uint count = NativeApi.Table->itemLoreCount(handle);
                List<string> lines = new List<string>((int)count);
                for (uint index = 0; index < count; index++)
                {
                    lines.Add(NativeApi.Text(NativeApi.Table->itemLore(handle, index)));
                }

                return lines;
            }
            set
            {
                SetLore(value);
            }
        }

        internal FalconItem* Raw
        {
            get
            {
                if (_handle == null)
                {
                    throw new ObjectDisposedException(nameof(Item));
                }

                return _handle;
            }
        }

        public static Item? Create(string identifier, uint count = 1)
        {
            fixed (byte* name = NativeApi.Utf8(identifier))
            {
                return Adopt(NativeApi.Table->itemCreate(name, count));
            }
        }

        public Item? Copy()
        {
            if (_handle == null)
            {
                return null;
            }

            return Adopt(NativeApi.Table->itemCopy(_handle));
        }

        public int GetEnchantmentLevel(uint enchantment)
        {
            return NativeApi.Table->itemEnchantmentLevel(Raw, enchantment);
        }

        public void SetEnchantmentLevel(uint enchantment, int level)
        {
            NativeApi.Table->itemSetEnchantmentLevel(Raw, enchantment, level);
        }

        public void Dispose()
        {
            if (_owned && _handle != null)
            {
                NativeApi.Table->itemDestroy(_handle);
            }

            _handle = null;
            _owned = false;
        }

        internal static Item? Adopt(FalconItem* handle)
        {
            if (handle == null)
            {
                return null;
            }

            return new Item(handle, true);
        }

        internal static Item Borrow(FalconItem* handle)
        {
            return new Item(handle, false);
        }

        private void SetLore(IReadOnlyList<string> lines)
        {
            FalconItem* handle = Raw;
            nint[] buffers = new nint[lines.Count];
            try
            {
                for (int index = 0; index < lines.Count; index++)
                {
                    buffers[index] = Marshal.StringToCoTaskMemUTF8(lines[index] ?? string.Empty);
                }

                fixed (nint* values = buffers)
                {
                    NativeApi.Table->itemSetLore(handle, (byte**)values, (uint)buffers.Length);
                }
            }
            finally
            {
                foreach (nint buffer in buffers)
                {
                    if (buffer != 0)
                    {
                        Marshal.FreeCoTaskMem(buffer);
                    }
                }
            }
        }
    }
}

File: java/src/main/java/falcon/api/Item.java
package falcon.api;

import falcon.api.internal.Interop;
import falcon.api.internal.NativeApi;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public final class Item implements AutoCloseable {
    private MemorySegment mHandle;
    private final boolean mOwned;

    private Item(MemorySegment handle, boolean owned) {
        mHandle = handle;
        mOwned = owned;
    }

    public static Optional<Item> create(String identifier) {
        return create(identifier, 1);
    }

    public static Optional<Item> create(String identifier, int count) {
        try (Arena arena = Arena.ofConfined()) {
            return adopt(api().itemCreate(Interop.text(arena, identifier), count));
        }
    }

    public static Optional<Item> adopt(MemorySegment handle) {
        if (Interop.isNull(handle)) {
            return Optional.empty();
        }
        return Optional.of(new Item(handle, true));
    }

    public static Item borrow(MemorySegment handle) {
        return new Item(handle == null ? MemorySegment.NULL : handle, false);
    }

    public Optional<Item> copy() {
        if (Interop.isNull(mHandle)) {
            return Optional.empty();
        }
        return adopt(api().itemCopy(mHandle));
    }

    @Override
    public void close() {
        if (!mOwned || Interop.isNull(mHandle)) {
            return;
        }
        api().itemDestroy(mHandle);
        mHandle = MemorySegment.NULL;
    }

    public boolean valid() {
        return !Interop.isNull(mHandle);
    }

    public boolean owned() {
        return mOwned;
    }

    public boolean isEmpty() {
        return Interop.isNull(mHandle) || api().itemIsEmpty(mHandle) != 0;
    }

    public String identifier() {
        return Interop.string(api().itemIdentifier(mHandle));
    }

    public int count() {
        return api().itemCount(mHandle);
    }

    public void setCount(int count) {
        api().itemSetCount(mHandle, count);
    }

    public int damage() {
        return api().itemDamage(mHandle);
    }

    public void setDamage(int damage) {
        api().itemSetDamage(mHandle, damage);
    }

    public String customName() {
        return Interop.string(api().itemCustomName(mHandle));
    }

    public void setCustomName(String name) {
        try (Arena arena = Arena.ofConfined()) {
            api().itemSetCustomName(mHandle, Interop.text(arena, name));
        }
    }

    public List<String> lore() {
        int count = api().itemLoreCount(mHandle);
        List<String> lines = new ArrayList<>(count);
        for (int index = 0; index < count; index++) {
            lines.add(Interop.string(api().itemLore(mHandle, index)));
        }
        return lines;
    }

    public void setLore(List<String> lines) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment values = MemorySegment.NULL;
            if (!lines.isEmpty()) {
                values = arena.allocate(ValueLayout.ADDRESS, lines.size());
                for (int index = 0; index < lines.size(); index++) {
                    values.setAtIndex(ValueLayout.ADDRESS, index, Interop.text(arena, lines.get(index)));
                }
            }
            api().itemSetLore(mHandle, values, lines.size());
        }
    }

    public int enchantmentLevel(int enchantment) {
        return api().itemEnchantmentLevel(mHandle, enchantment);
    }

    public void setEnchantmentLevel(int enchantment, int level) {
        api().itemSetEnchantmentLevel(mHandle, enchantment, level);
    }

    public MemorySegment handle() {
        return mHandle;
    }

    private static NativeApi api() {
        return Interop.api();
    }
}
